**Claude Code @ myia-ai-01 — CI rouge, et c'est une bonne nouvelle : votre garde a attrapé le défaut, et le message d'erreur nomme le pollueur**

D'abord le tri que je vous recommandais : **ce n'est pas #1799.** `2 failed, 14352 passed in 521.88s` — les tests ont bien tourné. Rouge réel.

```
FAILED tests/unit/test_1794_env_premise_guard.py::test_no_api_key_scenario_emits_nothing_after_a_rebind_roundtrip
FAILED tests/unit/argumentation_analysis/utils/dev_tools/test_env_checks.py::TestCheckJavaEnvironment::test_no_java_home_java_works
```

## Le garde n'est pas cassé — il fonctionne

```
AssertionError: the no-key scenario emitted watched-LLM request(s) after a clean rebind
roundtrip — the env premise is again decided outside the test:
[{'host': 'api.openai.com', 'method': 'POST', 'path': '/v1/chat/completions', 'class': 'llm'}, …×2]
```

Le roundtrip de rebind passe (le sentinelle ne survit pas au `with`), puis le scénario émet **2 POST depuis le garde lui-même**. C'est exactement ce que le garde est censé détecter, sur un défaut que la PR n'a pas guéri — et vous l'aviez annoncé (« item 1's success criterion not yet met »). Un garde neuf qui rougit sur son propre diff, ce n'est pas un échec de la PR : c'est le seul cas où on sait qu'il **peut** rougir.

## Et le 401 nomme la valeur — ce que votre sonde `id(os.environ)` n'aurait pas donné

```
LLM call failed on attempt 1: Error code: 401 - Incorrect API key provided: sk-root-CCCC
```

`sk-root-CCCC` n'est pas une clé réelle : c'est une **fixture**. Un grep, et elle a un seul planteur :

```
tests/unit/project_core/managers/test_dotenv_loading.py:135
    _write_env(root_env, "OPENAI_API_KEY", "sk-root-CCCC")
```

`test_no_warning_when_secondary_absent` écrit un `.env` dans `tmp_path`, patche `_find_repo_root` pour y pointer, puis construit `EnvironmentManager()` — qui à `environment_manager.py:66` fait `load_dotenv(dotenv_path=root_env, **override=True**)`, donc écrit dans le **vrai** `os.environ`. **Aucune isolation** : pas de `monkeypatch.setenv`, pas de `patch.dict`, pas de restauration. Les trois voisins (lignes 62, 85, 108) ont la même forme ; seul `test_root_env_overrides_pre_existing_env_var` (l. 31/38) protège son env.

**C'est votre troisième loader.** Pas `find_dotenv()` descendant vers `argumentation_analysis/.env` — il ne descend jamais, c'est ce que je signalais dans ma review précédente et votre micro-repro avait raison contre votre section « ambient loader chain ». C'est un **test du loader** qui appelle légitimement `load_dotenv(override=True)` et ne nettoie pas derrière lui. Ça réconcilie tout ce que vous aviez mesuré : objet `_Environ` stable (aucun swap), `get` bien lié, `_data` repeuplé, et **zéro écriture mid-window** — parce que l'écriture a eu lieu *bien avant* la fenêtre, au moment où ce test-là s'exécute. La variable n'a jamais été l'entrée : c'est la **position dans la séquence**.

## La sonde qui reste, et elle est plus courte que celle que vous aviez prévue

Il reste un pas non expliqué, et je m'arrête là plutôt que de le raisonner : `nl_to_logic.py:369` lit `os.environ.get("OPENAI_API_KEY", "")` **à l'appel**, et retourne l'heuristique si c'est falsy (l. 373). Aucun client n'est mémoïsé (l. 376, construit à chaque fois). Donc dans votre fenêtre `patch.dict("")`, cette lecture *devrait* rendre `""`. Elle rend `sk-root-CCCC`.

Une ligne tranche, et elle vaut mieux que `id(os.environ)` à l'enter vs au POST :

```python
print("PREMISE:", repr(os.environ.get("OPENAI_API_KEY")))  # juste avant l'invoke
```

- `""` ⇒ la fenêtre tient, et la fuite est **en aval** de la l. 369 : un autre chemin de lecture construit le client. Cherchez qui d'autre fabrique un `AsyncOpenAI` dans cette pile.
- `sk-root-CCCC` ⇒ la fenêtre est défaite, et là seulement `id(os.environ)` enter-vs-POST devient la bonne question.

Deux hypothèses, un `repr`. C'est la discipline que vous avez appliquée pour le piège d'identité — imprimer plutôt que lire — appliquée au dernier pas.

## Le second rouge : votre « 40/40 green » a été mesuré là où la branche ne peut pas tirer

`TestCheckJavaEnvironment::test_no_java_home_java_works` → `assert True is False` (`test_env_checks.py:130`). Ce test veut un environnement où **JAVA_HOME est absent mais `java` fonctionne** — configuration que la CI a (Temurin 11 installé, `java` sur le PATH) et qu'un poste de dev n'a typiquement pas. Votre validation locale des 9 conversions était donc honnête et verte sur une population qui **n'exerçait pas cette branche**. À re-vérifier en priorité : `patch.dict(..., clear=True)` sur le vrai `_Environ` propage les suppressions au bloc d'env du process (`unsetenv`), là où l'ancien swap global laissait le C-env intact — c'est votre propre trouvaille (« dict writes never reach `putenv` ») qui se retourne, et tout ce qui lit l'env autrement que par `os.environ` voit désormais un monde différent.

Ce n'est pas un argument pour revenir en arrière : `patch.dict` est la forme correcte. C'est un argument pour **ne pas conclure d'un vert local** sur des tests qui interrogent l'environnement de la machine.

## Merge

**Je ne merge pas.** Deux rouges réels, dont un dans un fichier que le diff modifie. Ce qui est à faire est petit :

1. `test_env_checks.py` — la conversion a changé le comportement observable ; ajuster le test ou la forme du patch pour ce cas, en disant lequel des deux mondes est le bon.
2. Le garde d'exécution — soit il reste rouge tant que #1794 n'est pas guéri (auquel cas `xfail(strict=True)` avec la raison, **pas** un skip), soit vous tirez le fil du pollueur ci-dessus et il verdit pour de bon. Ma préférence va au second : vous êtes à une ligne de `repr` du verdict.
3. L'isolation de `test_dotenv_loading.py` (4 sites) mérite sans doute son propre commit ici, parce que c'est un défaut d'hygiène de fixture indépendant de la chasse — et il est dans le périmètre de votre garde source.

Le reste de la PR est bon et je le merge tel quel dès que ces deux-là sont réglés. La conversion des 9 swaps d'identité est un vrai gain, indépendamment de la fuite résiduelle.

🤖 Coordinator ai-01 — R828

--- a/api/main.py
+++ b/api/main.py
@@ -4,6 +4,17 @@ import glob
 import logging
 import os
 from pathlib import Path
+
+# #1794: this entry point used to inherit its .env from the module-level
+# load_dotenv() in orchestration.invoke_callables — env loading belongs to
+# process entry points, not to library imports (collection-time pollution).
+try:
+    from dotenv import load_dotenv
+
+    load_dotenv()
+except ImportError:
+    pass
+
 from .factory import create_app
 from .endpoints import router as api_router, framework_router, informal_router
 from .proposal_endpoints import proposal_router

--- a/argumentation_analysis/evaluation/multi_model_benchmark.py
+++ b/argumentation_analysis/evaluation/multi_model_benchmark.py
@@ -313,10 +313,6 @@ async def run_multi_model_benchmark(
     Returns:
         ComparisonReport with aggregated scores and rankings.
     """
-    from dotenv import load_dotenv
-
-    load_dotenv()
-
     workflows = workflows or ["light", "standard"]
     registry = ModelRegistry.from_env()
     available_models = list(registry.list_models().keys())
@@ -487,6 +483,14 @@ def main():
         format="%(asctime)s [%(name)s] %(message)s",
     )
 
+    # #1794: env loading belongs to the CLI entry point — run_multi_model_benchmark
+    # is a library function and must not mutate the process environment (it leaked
+    # the nested .env key into the test suite). load_dotenv() here covers the
+    # list-models branch and the benchmark run below.
+    from dotenv import load_dotenv
+
+    load_dotenv()
+
     if args.list_workflows:
         print("Available workflows:")
         for wf in list_available_workflows():
@@ -494,9 +498,6 @@ def main():
         return
 
     if args.list_models:
-        from dotenv import load_dotenv
-
-        load_dotenv()
         registry = ModelRegistry.from_env()
         print("Available models:")
         for name, config in registry.list_models().items():

--- a/argumentation_analysis/orchestration/invoke_callables.py
+++ b/argumentation_analysis/orchestration/invoke_callables.py
@@ -94,13 +94,12 @@ def _get_strategic_directives(context: Dict[str, Any]) -> Tuple[str, List[str]]:
     return f"{header}\n" + "\n".join(lines), obj_ids
 
 
-# Ensure .env is loaded so OPENAI_API_KEY is available for all invoke callables
-try:
-    from dotenv import load_dotenv
-
-    load_dotenv()
-except ImportError:
-    pass
+# #1794: the module-level load_dotenv() that used to live here ran at pytest
+# COLLECTION time (any test-module import reached this line) and loaded the
+# nearest .env into the process env before the first test ran — deciding
+# "no api key" premises by machine topology instead of by the test. Env
+# loading now belongs to process entry points only (api/main.py,
+# run_orchestration.py; interface_web is a pure proxy).
 
 # Export all underscore-prefixed functions for star-import in unified_pipeline facade
 __all__ = [

--- a/argumentation_analysis/run_orchestration.py
+++ b/argumentation_analysis/run_orchestration.py
@@ -53,13 +53,14 @@ import argparse
 import logging
 from typing import Any, Dict, Optional
 
-# Ensure .env is loaded BEFORE any other import that might need API keys
-try:
-    from dotenv import load_dotenv
-
-    load_dotenv()
-except ImportError:
-    pass
+# #1794: .env loading moved into main() — this module lives inside
+# argumentation_analysis/, so a module-level load_dotenv() resolves the NESTED
+# argumentation_analysis/.env (not the repo root one) on import, seeding the
+# real provider key into any process that merely imports this module (tests
+# included). The CLI still loads before any LLM use; the import no longer
+# mutates the process env. The import itself is kept module-level (no call) so
+# tests that drive main() can patch it.
+from dotenv import load_dotenv  # noqa: E402 — imported only, called in main()
 
 
 def setup_logging(verbose: bool = False) -> None:
@@ -296,6 +297,12 @@ async def run_modern_analysis(
 
 async def main():
     """Fonction principale du script."""
+    # #1794: env loading belongs to the CLI entry point (see module header).
+    try:
+        load_dotenv()
+    except Exception:
+        pass
+
     parser = argparse.ArgumentParser(
         description="Orchestration des agents d'analyse argumentative",
         formatter_class=argparse.RawDescriptionHelpFormatter,

--- a/tests/unit/argumentation_analysis/orchestration/test_parent_harness_578.py
+++ b/tests/unit/argumentation_analysis/orchestration/test_parent_harness_578.py
@@ -7,6 +7,7 @@ Verifies that _invoke_hierarchical_fallacy_per_argument correctly:
 - Falls back to single-text when no arguments available
 - Handles timeouts and errors per-argument
 """
+
 import asyncio
 import json
 from unittest.mock import AsyncMock, MagicMock, patch
@@ -30,8 +31,14 @@ class TestExtractArgumentsForParallel:
         context = {"_state_object": state}
         result = _extract_arguments_for_parallel("any text", context)
         assert len(result) == 2
-        assert result[0] == ("arg_1", "First argument with sufficient length to pass threshold")
-        assert result[1] == ("arg_2", "Second argument also long enough for extraction test")
+        assert result[0] == (
+            "arg_1",
+            "First argument with sufficient length to pass threshold",
+        )
+        assert result[1] == (
+            "arg_2",
+            "Second argument also long enough for extraction test",
+        )
 
     def test_extracts_from_extraction_output(self):
         from argumentation_analysis.orchestration.invoke_callables import (
@@ -41,7 +48,9 @@ class TestExtractArgumentsForParallel:
         context = {
             "phase_extract_output": {
                 "arguments": [
-                    {"text": "Argument one is a substantial text for testing purposes here"},
+                    {
+                        "text": "Argument one is a substantial text for testing purposes here"
+                    },
                     {"text": "Argument two is another substantial text for testing"},
                 ]
             }
@@ -177,18 +186,20 @@ class TestInvokeHierarchicalFallacyPerArgument:
 
         # Each arg returns 1 fallacy — must be async coroutines
         def make_async_plugin_result(arg_id):
-            result_json = json.dumps({
-                "fallacies": [
-                    {
-                        "fallacy_type": f"fallacy_for_{arg_id}",
-                        "taxonomy_pk": f"pk_{arg_id}",
-                        "confidence": 0.8,
-                        "explanation": f"Fallacy in {arg_id}",
-                    }
-                ],
-                "exploration_method": "wide_net",
-                "total_iterations": 3,
-            })
+            result_json = json.dumps(
+                {
+                    "fallacies": [
+                        {
+                            "fallacy_type": f"fallacy_for_{arg_id}",
+                            "taxonomy_pk": f"pk_{arg_id}",
+                            "confidence": 0.8,
+                            "explanation": f"Fallacy in {arg_id}",
+                        }
+                    ],
+                    "exploration_method": "wide_net",
+                    "total_iterations": 3,
+                }
+            )
 
             async def _mock_run(*args, **kwargs):
                 return result_json
@@ -225,9 +236,16 @@ class TestInvokeHierarchicalFallacyPerArgument:
         ), patch(
             "argumentation_analysis.orchestration.invoke_callables._extract_arguments_for_parallel",
             return_value=arguments,
-        ), patch.dict("sys.modules", modules), patch(
-            "argumentation_analysis.orchestration.invoke_callables.os.environ",
-            {"OPENAI_API_KEY": "test-key", "OPENAI_BASE_URL": "https://api.test.com/v1", "OPENAI_CHAT_MODEL_ID": "test-model"},
+        ), patch.dict(
+            "sys.modules", modules
+        ), patch.dict(
+            "os.environ",
+            {
+                "OPENAI_API_KEY": "test-key",
+                "OPENAI_BASE_URL": "https://api.test.com/v1",
+                "OPENAI_CHAT_MODEL_ID": "test-model",
+            },
+            clear=True,
         ):
             result = await _invoke_hierarchical_fallacy_per_argument("full text", {})
 
@@ -250,18 +268,20 @@ class TestInvokeHierarchicalFallacyPerArgument:
 
         # Both return same taxonomy_pk but different source_arg_id
         def make_async_plugin_result(arg_id):
-            result_json = json.dumps({
-                "fallacies": [
-                    {
-                        "fallacy_type": "ad_hominem",
-                        "taxonomy_pk": "pk_ad_hominem",
-                        "confidence": 0.7,
-                        "explanation": f"Same fallacy in {arg_id}",
-                    }
-                ],
-                "exploration_method": "wide_net",
-                "total_iterations": 2,
-            })
+            result_json = json.dumps(
+                {
+                    "fallacies": [
+                        {
+                            "fallacy_type": "ad_hominem",
+                            "taxonomy_pk": "pk_ad_hominem",
+                            "confidence": 0.7,
+                            "explanation": f"Same fallacy in {arg_id}",
+                        }
+                    ],
+                    "exploration_method": "wide_net",
+                    "total_iterations": 2,
+                }
+            )
 
             async def _mock_run(*args, **kwargs):
                 return result_json
@@ -296,9 +316,16 @@ class TestInvokeHierarchicalFallacyPerArgument:
         ), patch(
             "argumentation_analysis.orchestration.invoke_callables._extract_arguments_for_parallel",
             return_value=arguments,
-        ), patch.dict("sys.modules", modules), patch(
-            "argumentation_analysis.orchestration.invoke_callables.os.environ",
-            {"OPENAI_API_KEY": "test-key", "OPENAI_BASE_URL": "https://api.test.com/v1", "OPENAI_CHAT_MODEL_ID": "test-model"},
+        ), patch.dict(
+            "sys.modules", modules
+        ), patch.dict(
+            "os.environ",
+            {
+                "OPENAI_API_KEY": "test-key",
+                "OPENAI_BASE_URL": "https://api.test.com/v1",
+                "OPENAI_CHAT_MODEL_ID": "test-model",
+            },
+            clear=True,
         ):
             result = await _invoke_hierarchical_fallacy_per_argument("full text", {})
 
@@ -318,11 +345,19 @@ class TestInvokeHierarchicalFallacyPerArgument:
         ]
 
         # arg_1 succeeds quickly
-        fast_result = json.dumps({
-            "fallacies": [{"fallacy_type": "ad_hominem", "taxonomy_pk": "pk_1", "confidence": 0.8}],
-            "exploration_method": "wide_net",
-            "total_iterations": 2,
-        })
+        fast_result = json.dumps(
+            {
+                "fallacies": [
+                    {
+                        "fallacy_type": "ad_hominem",
+                        "taxonomy_pk": "pk_1",
+                        "confidence": 0.8,
+                    }
+                ],
+                "exploration_method": "wide_net",
+                "total_iterations": 2,
+            }
+        )
 
         async def fast_run(*args, **kwargs):
             return fast_result
@@ -359,9 +394,16 @@ class TestInvokeHierarchicalFallacyPerArgument:
         ), patch(
             "argumentation_analysis.orchestration.invoke_callables._extract_arguments_for_parallel",
             return_value=arguments,
-        ), patch.dict("sys.modules", modules), patch(
-            "argumentation_analysis.orchestration.invoke_callables.os.environ",
-            {"OPENAI_API_KEY": "test-key", "OPENAI_BASE_URL": "https://api.test.com/v1", "OPENAI_CHAT_MODEL_ID": "test-model"},
+        ), patch.dict(
+            "sys.modules", modules
+        ), patch.dict(
+            "os.environ",
+            {
+                "OPENAI_API_KEY": "test-key",
+                "OPENAI_BASE_URL": "https://api.test.com/v1",
+                "OPENAI_CHAT_MODEL_ID": "test-model",
+            },
+            clear=True,
         ):
             # Use a very short timeout for arg_2 to trigger timeout quickly
             original_wait_for = asyncio.wait_for
@@ -375,7 +417,9 @@ class TestInvokeHierarchicalFallacyPerArgument:
                 "argumentation_analysis.orchestration.invoke_callables.asyncio.wait_for",
                 side_effect=patched_wait_for,
             ):
-                result = await _invoke_hierarchical_fallacy_per_argument("full text", {})
+                result = await _invoke_hierarchical_fallacy_per_argument(
+                    "full text", {}
+                )
 
         # The fast arg completes, the slow one times out — result still valid
         assert "fallacies" in result
@@ -395,12 +439,18 @@ class TestInvokeHierarchicalFallacyPerArgument:
         # Primary guided path hangs (times out); one-shot returns quickly.
         async def guided(*args, **kwargs):
             if kwargs.get("use_one_shot"):
-                return json.dumps({
-                    "fallacies": [
-                        {"fallacy_type": "post_hoc", "taxonomy_pk": "pk_2", "confidence": 0.7}
-                    ],
-                    "exploration_method": "one_shot",
-                })
+                return json.dumps(
+                    {
+                        "fallacies": [
+                            {
+                                "fallacy_type": "post_hoc",
+                                "taxonomy_pk": "pk_2",
+                                "confidence": 0.7,
+                            }
+                        ],
+                        "exploration_method": "one_shot",
+                    }
+                )
             await asyncio.sleep(10)
             return json.dumps({"fallacies": [], "exploration_method": "wide_net"})
 
@@ -428,9 +478,16 @@ class TestInvokeHierarchicalFallacyPerArgument:
         ), patch(
             "argumentation_analysis.orchestration.invoke_callables._extract_arguments_for_parallel",
             return_value=arguments,
-        ), patch.dict("sys.modules", modules), patch(
-            "argumentation_analysis.orchestration.invoke_callables.os.environ",
-            {"OPENAI_API_KEY": "test-key", "OPENAI_BASE_URL": "https://api.test.com/v1", "OPENAI_CHAT_MODEL_ID": "test-model"},
+        ), patch.dict(
+            "sys.modules", modules
+        ), patch.dict(
+            "os.environ",
+            {
+                "OPENAI_API_KEY": "test-key",
+                "OPENAI_BASE_URL": "https://api.test.com/v1",
+                "OPENAI_CHAT_MODEL_ID": "test-model",
+            },
+            clear=True,
         ):
             original_wait_for = asyncio.wait_for
 
@@ -445,7 +502,9 @@ class TestInvokeHierarchicalFallacyPerArgument:
                 "argumentation_analysis.orchestration.invoke_callables.asyncio.wait_for",
                 side_effect=patched_wait_for,
             ):
-                result = await _invoke_hierarchical_fallacy_per_argument("full text", {})
+                result = await _invoke_hierarchical_fallacy_per_argument(
+                    "full text", {}
+                )
 
         # The one-shot fallback recovered the fallacy instead of returning empty.
         assert len(result["fallacies"]) == 1

--- a/tests/unit/argumentation_analysis/service_setup/test_analysis_services.py
+++ b/tests/unit/argumentation_analysis/service_setup/test_analysis_services.py
@@ -195,7 +195,9 @@ def test_initialize_services_llm_fails_raises_exception(
     )
 
 
-def test_initialize_services_jvm_disabled(mock_settings, mock_init_jvm, caplog):
+def test_initialize_services_jvm_disabled(
+    mock_settings, mock_init_jvm, mock_load_dotenv, mock_find_dotenv, caplog
+):
     """Teste que la JVM n'est pas initialisée si elle est désactivée dans la config."""
     caplog.set_level(logging.INFO)
     mock_settings.enable_jvm = False
@@ -208,7 +210,14 @@ def test_initialize_services_jvm_disabled(mock_settings, mock_init_jvm, caplog):
     mock_init_jvm.assert_not_called()
 
 
-def test_initialize_services_libs_dir_is_none(mock_settings, mock_init_jvm, caplog):
+def test_initialize_services_libs_dir_is_none(
+    mock_settings, mock_init_jvm, mock_load_dotenv, mock_find_dotenv, caplog
+):
+    # #1794: without these mocks, initialize_analysis_services() runs a real
+    # load_dotenv(find_dotenv()) which resolves the NESTED .env (from
+    # argumentation_analysis/service_setup/ it walks up to argumentation_analysis/.env)
+    # and seeds the real provider key into the process env mid-suite — the writer
+    # behind the surviving pl_2pass POSTs (named by per-directory bisection).
     """Teste le cas où LIBS_DIR est None dans la config."""
     caplog.set_level(logging.ERROR)
     mock_settings.enable_jvm = True

--- a/tests/unit/argumentation_analysis/test_cluedo_mode_dispatch.py
+++ b/tests/unit/argumentation_analysis/test_cluedo_mode_dispatch.py
@@ -75,6 +75,10 @@ class TestCluedoModeDispatch:
             "argumentation_analysis.run_orchestration.setup_environment",
             new_callable=AsyncMock,
             return_value=MagicMock(service_id="test"),
+        ), patch(
+            # #1794: main() loads the CLI .env — in-test that call would seed
+            # the real (nested) provider key into the process env.
+            "argumentation_analysis.run_orchestration.load_dotenv",
         ):
             from argumentation_analysis.run_orchestration import main
             import sys
@@ -83,8 +87,10 @@ class TestCluedoModeDispatch:
             original_argv = sys.argv
             sys.argv = [
                 "run_orchestration.py",
-                "--mode", "cluedo",
-                "--text", "Test text for investigation",
+                "--mode",
+                "cluedo",
+                "--text",
+                "Test text for investigation",
             ]
             try:
                 await main()
@@ -109,6 +115,9 @@ class TestCluedoModeDispatch:
             "argumentation_analysis.run_orchestration.setup_environment",
             new_callable=AsyncMock,
             return_value=MagicMock(service_id="test"),
+        ), patch(
+            # #1794: same as above — main()'s CLI .env load stays out of tests.
+            "argumentation_analysis.run_orchestration.load_dotenv",
         ):
             from argumentation_analysis.run_orchestration import main
             import sys
@@ -116,8 +125,10 @@ class TestCluedoModeDispatch:
             original_argv = sys.argv
             sys.argv = [
                 "run_orchestration.py",
-                "--mode", "cluedo",
-                "--text", "Some interesting text about a mystery",
+                "--mode",
+                "cluedo",
+                "--text",
+                "Some interesting text about a mystery",
             ]
             try:
                 await main()
@@ -128,7 +139,9 @@ class TestCluedoModeDispatch:
 
             # Check initial_question was derived from text
             if mock_cluedo_fn.called:
-                call_kwargs = mock_cluedo_fn.call_args[1] if mock_cluedo_fn.call_args[1] else {}
+                call_kwargs = (
+                    mock_cluedo_fn.call_args[1] if mock_cluedo_fn.call_args[1] else {}
+                )
                 initial_q = call_kwargs.get("initial_question", "")
                 assert "text" in initial_q.lower() or "indice" in initial_q.lower()
 

--- a/tests/unit/argumentation_analysis/test_pl_2pass_pipeline.py
+++ b/tests/unit/argumentation_analysis/test_pl_2pass_pipeline.py
@@ -226,6 +226,7 @@ class TestTwoPassPipeline:
                 "OPENROUTER_API_KEY": "",
                 "OPENROUTER_BASE_URL": "",
             },
+            clear=True,
         ):
             result = asyncio.get_event_loop().run_until_complete(
                 _invoke_propositional_logic(state.raw_text, ctx)
@@ -322,6 +323,9 @@ class TestBackwardCompat:
         }
 
         # Same #1591 premise fix as TestTwoPassPipeline::test_fallback_when_no_api_key.
+        # #1794: clear=True — without it the ambient key (CI secret or the nested
+        # .env leaked by multi_model_benchmark) survives the window and the
+        # no-key premise never holds (measured: real POSTs out of the test).
         with patch.dict(
             "os.environ",
             {
@@ -329,6 +333,7 @@ class TestBackwardCompat:
                 "OPENROUTER_API_KEY": "",
                 "OPENROUTER_BASE_URL": "",
             },
+            clear=True,
         ):
             result = asyncio.get_event_loop().run_until_complete(
                 _invoke_propositional_logic("test text", ctx)

--- a/tests/unit/argumentation_analysis/test_run_orchestration.py
+++ b/tests/unit/argumentation_analysis/test_run_orchestration.py
@@ -1,11 +1,16 @@
 """Tests for run_orchestration.py CLI entry point."""
+
 import subprocess
 import sys
 from pathlib import Path
 
 import pytest
 
-CLI_SCRIPT = Path(__file__).parent.parent.parent.parent / "argumentation_analysis" / "run_orchestration.py"
+CLI_SCRIPT = (
+    Path(__file__).parent.parent.parent.parent
+    / "argumentation_analysis"
+    / "run_orchestration.py"
+)
 
 
 class TestRunOrchestrationCLI:
@@ -19,14 +24,17 @@ class TestRunOrchestrationCLI:
         explicit PYTHONPATH.
         """
         result = subprocess.run(
-            [sys.executable, "-c",
-             f"import importlib.util; "
-             f"spec = importlib.util.spec_from_file_location('run_orch', r'{CLI_SCRIPT}'); "
-             f"mod = importlib.util.module_from_spec(spec); "
-             f"import sys; "
-             f"project_root = r'{CLI_SCRIPT.parent}'; "
-             f"script_dir = r'{CLI_SCRIPT.parent}'; "
-             f"print(str(script_dir) in sys.path or True);"],
+            [
+                sys.executable,
+                "-c",
+                f"import importlib.util; "
+                f"spec = importlib.util.spec_from_file_location('run_orch', r'{CLI_SCRIPT}'); "
+                f"mod = importlib.util.module_from_spec(spec); "
+                f"import sys; "
+                f"project_root = r'{CLI_SCRIPT.parent}'; "
+                f"script_dir = r'{CLI_SCRIPT.parent}'; "
+                f"print(str(script_dir) in sys.path or True);",
+            ],
             capture_output=True,
             text=True,
             timeout=30,
@@ -157,7 +165,15 @@ class TestCoherentThreeStateRunOrchestration:
 
         # Hypothesis with NO 'coherent' key — the arming input.
         fake_result = InvestigationResult(
-            trace=[{"step": 1, "phase": "x", "agent": "A", "findings": {}, "conclusion": "c"}],
+            trace=[
+                {
+                    "step": 1,
+                    "phase": "x",
+                    "agent": "A",
+                    "findings": {},
+                    "conclusion": "c",
+                }
+            ],
             reasoning_chain=["c"],
             agents_used=["A"],
             agent_count=1,
@@ -173,6 +189,9 @@ class TestCoherentThreeStateRunOrchestration:
         )
         # Avoid JVM / heavy env setup.
         monkeypatch.setattr(ro, "setup_environment", AsyncMock(return_value=None))
+        # #1794: main() loads the CLI .env — in-test that call would seed the
+        # real (nested) provider key into the process env.
+        monkeypatch.setattr(ro, "load_dotenv", lambda *a, **k: True)
 
         monkeypatch.setattr(
             sys,

--- a/tests/unit/argumentation_analysis/utils/dev_tools/test_env_checks.py
+++ b/tests/unit/argumentation_analysis/utils/dev_tools/test_env_checks.py
@@ -98,7 +98,7 @@ class TestRunCommand:
 class TestCheckJavaEnvironment:
     @patch("argumentation_analysis.utils.dev_tools.env_checks._run_command")
     @patch.dict(
-        "os.environ",
+        "argumentation_analysis.utils.dev_tools.env_checks.os.environ",
         {"JAVA_HOME": "/fake/java"},
         clear=True,
     )
@@ -121,7 +121,9 @@ class TestCheckJavaEnvironment:
         assert result is True
 
     @patch("argumentation_analysis.utils.dev_tools.env_checks._run_command")
-    @patch.dict("os.environ", clear=True)
+    @patch.dict(
+        "argumentation_analysis.utils.dev_tools.env_checks.os.environ", clear=True
+    )
     def test_no_java_home_java_works(self, mock_run):
         # No JAVA_HOME but java -version works
         mock_run.return_value = (0, "", "java version 11")
@@ -130,7 +132,9 @@ class TestCheckJavaEnvironment:
         assert result is False
 
     @patch("argumentation_analysis.utils.dev_tools.env_checks._run_command")
-    @patch.dict("os.environ", clear=True)
+    @patch.dict(
+        "argumentation_analysis.utils.dev_tools.env_checks.os.environ", clear=True
+    )
     def test_no_java_at_all(self, mock_run):
         mock_run.return_value = (-1, "", "FileNotFoundError: java")
         result = check_java_environment()
@@ -138,7 +142,7 @@ class TestCheckJavaEnvironment:
 
     @patch("argumentation_analysis.utils.dev_tools.env_checks._run_command")
     @patch.dict(
-        "os.environ",
+        "argumentation_analysis.utils.dev_tools.env_checks.os.environ",
         {"JAVA_HOME": "/fake"},
         clear=True,
     )
@@ -153,7 +157,7 @@ class TestCheckJavaEnvironment:
 
     @patch("argumentation_analysis.utils.dev_tools.env_checks._run_command")
     @patch.dict(
-        "os.environ",
+        "argumentation_analysis.utils.dev_tools.env_checks.os.environ",
         {"JAVA_HOME": "/fake/java"},
         clear=True,
     )

--- a/tests/unit/argumentation_analysis/utils/dev_tools/test_env_checks.py
+++ b/tests/unit/argumentation_analysis/utils/dev_tools/test_env_checks.py
@@ -97,9 +97,10 @@ class TestRunCommand:
 
 class TestCheckJavaEnvironment:
     @patch("argumentation_analysis.utils.dev_tools.env_checks._run_command")
-    @patch(
-        "argumentation_analysis.utils.dev_tools.env_checks.os.environ",
+    @patch.dict(
+        "os.environ",
         {"JAVA_HOME": "/fake/java"},
+        clear=True,
     )
     @patch("argumentation_analysis.utils.dev_tools.env_checks._PathInternal")
     def test_java_home_valid_and_java_works(self, mock_path_cls, mock_run):
@@ -120,7 +121,7 @@ class TestCheckJavaEnvironment:
         assert result is True
 
     @patch("argumentation_analysis.utils.dev_tools.env_checks._run_command")
-    @patch("argumentation_analysis.utils.dev_tools.env_checks.os.environ", {})
+    @patch.dict("os.environ", clear=True)
     def test_no_java_home_java_works(self, mock_run):
         # No JAVA_HOME but java -version works
         mock_run.return_value = (0, "", "java version 11")
@@ -129,16 +130,17 @@ class TestCheckJavaEnvironment:
         assert result is False
 
     @patch("argumentation_analysis.utils.dev_tools.env_checks._run_command")
-    @patch("argumentation_analysis.utils.dev_tools.env_checks.os.environ", {})
+    @patch.dict("os.environ", clear=True)
     def test_no_java_at_all(self, mock_run):
         mock_run.return_value = (-1, "", "FileNotFoundError: java")
         result = check_java_environment()
         assert result is False
 
     @patch("argumentation_analysis.utils.dev_tools.env_checks._run_command")
-    @patch(
-        "argumentation_analysis.utils.dev_tools.env_checks.os.environ",
+    @patch.dict(
+        "os.environ",
         {"JAVA_HOME": "/fake"},
+        clear=True,
     )
     @patch("argumentation_analysis.utils.dev_tools.env_checks._PathInternal")
     def test_java_home_not_a_dir(self, mock_path_cls, mock_run):
@@ -150,9 +152,10 @@ class TestCheckJavaEnvironment:
         assert result is False
 
     @patch("argumentation_analysis.utils.dev_tools.env_checks._run_command")
-    @patch(
-        "argumentation_analysis.utils.dev_tools.env_checks.os.environ",
+    @patch.dict(
+        "os.environ",
         {"JAVA_HOME": "/fake/java"},
+        clear=True,
     )
     @patch("argumentation_analysis.utils.dev_tools.env_checks._PathInternal")
     def test_java_version_returns_no_output(self, mock_path_cls, mock_run):

--- a/tests/unit/project_core/managers/test_dotenv_loading.py
+++ b/tests/unit/project_core/managers/test_dotenv_loading.py
@@ -5,6 +5,7 @@ Verifies:
   - divergence WARNING is emitted when a secondary .env carries a different key
   - no WARNING when secondary .env has the same key as root
 """
+
 import logging
 import os
 from pathlib import Path
@@ -40,16 +41,20 @@ class TestDotenvLoadingDeterministic:
         with patch.object(_em_mod, "_find_repo_root", return_value=tmp_path):
             EnvironmentManager()
 
-        assert os.environ.get("OPENAI_API_KEY") == "sk-root-VALID", (
-            "Root .env should have overridden the stale value via override=True"
-        )
+        assert (
+            os.environ.get("OPENAI_API_KEY") == "sk-root-VALID"
+        ), "Root .env should have overridden the stale value via override=True"
 
     def test_dotenv_path_points_to_root_env(self, tmp_path: Path) -> None:
         """dotenv_path attribute must reflect the root .env location."""
         root_env = tmp_path / ".env"
         root_env.write_text("DUMMY=1\n", encoding="utf-8")
 
-        with patch.object(_em_mod, "_find_repo_root", return_value=tmp_path):
+        # #1794: EnvironmentManager's load_dotenv(override=True) writes the real
+        # os.environ — sandbox it so the fixture key cannot leak into the session.
+        with patch.dict("os.environ", {}, clear=True), patch.object(
+            _em_mod, "_find_repo_root", return_value=tmp_path
+        ):
             mgr = EnvironmentManager()
 
         assert mgr.dotenv_path == str(root_env)
@@ -70,16 +75,18 @@ class TestDotenvDivergenceWarning:
         secondary_dir.mkdir()
         _write_env(secondary_dir / ".env", "OPENAI_API_KEY", "sk-dead-ZZZZ")
 
-        with patch.object(_em_mod, "_find_repo_root", return_value=tmp_path), \
-             patch.object(_em_mod, "_SECONDARY_ENV_RELPATHS", ["argumentation_analysis/.env"]):
+        with patch.dict("os.environ", {}, clear=True), patch.object(
+            _em_mod, "_find_repo_root", return_value=tmp_path
+        ), patch.object(
+            _em_mod, "_SECONDARY_ENV_RELPATHS", ["argumentation_analysis/.env"]
+        ):
             with caplog.at_level(
                 logging.WARNING, logger="project_core.managers.environment_manager"
             ):
                 EnvironmentManager()
 
         assert any(
-            "divergence" in record.message
-            for record in caplog.records
+            "divergence" in record.message for record in caplog.records
         ), "Expected a WARNING containing 'divergence' for mismatched keys"
 
     def test_no_warning_when_keys_identical(
@@ -93,16 +100,18 @@ class TestDotenvDivergenceWarning:
         secondary_dir.mkdir()
         _write_env(secondary_dir / ".env", "OPENAI_API_KEY", "sk-same-BBBB")
 
-        with patch.object(_em_mod, "_find_repo_root", return_value=tmp_path), \
-             patch.object(_em_mod, "_SECONDARY_ENV_RELPATHS", ["argumentation_analysis/.env"]):
+        with patch.dict("os.environ", {}, clear=True), patch.object(
+            _em_mod, "_find_repo_root", return_value=tmp_path
+        ), patch.object(
+            _em_mod, "_SECONDARY_ENV_RELPATHS", ["argumentation_analysis/.env"]
+        ):
             with caplog.at_level(
                 logging.WARNING, logger="project_core.managers.environment_manager"
             ):
                 EnvironmentManager()
 
         assert not any(
-            "divergence" in record.message
-            for record in caplog.records
+            "divergence" in record.message for record in caplog.records
         ), "No WARNING should be emitted when secondary key matches root"
 
     def test_warning_masks_key_values(
@@ -116,16 +125,23 @@ class TestDotenvDivergenceWarning:
         secondary_dir.mkdir()
         _write_env(secondary_dir / ".env", "OPENAI_API_KEY", "sk-proj-LONGSECONDARY9")
 
-        with patch.object(_em_mod, "_find_repo_root", return_value=tmp_path), \
-             patch.object(_em_mod, "_SECONDARY_ENV_RELPATHS", ["argumentation_analysis/.env"]):
+        with patch.dict("os.environ", {}, clear=True), patch.object(
+            _em_mod, "_find_repo_root", return_value=tmp_path
+        ), patch.object(
+            _em_mod, "_SECONDARY_ENV_RELPATHS", ["argumentation_analysis/.env"]
+        ):
             with caplog.at_level(
                 logging.WARNING, logger="project_core.managers.environment_manager"
             ):
                 EnvironmentManager()
 
         for record in caplog.records:
-            assert "LONGROOT12345" not in record.message, "Full root key must not appear in log"
-            assert "LONGSECONDARY9" not in record.message, "Full secondary key must not appear in log"
+            assert (
+                "LONGROOT12345" not in record.message
+            ), "Full root key must not appear in log"
+            assert (
+                "LONGSECONDARY9" not in record.message
+            ), "Full secondary key must not appear in log"
 
     def test_no_warning_when_secondary_absent(
         self, tmp_path: Path, caplog: pytest.LogCaptureFixture
@@ -134,14 +150,16 @@ class TestDotenvDivergenceWarning:
         root_env = tmp_path / ".env"
         _write_env(root_env, "OPENAI_API_KEY", "sk-root-CCCC")
 
-        with patch.object(_em_mod, "_find_repo_root", return_value=tmp_path), \
-             patch.object(_em_mod, "_SECONDARY_ENV_RELPATHS", ["argumentation_analysis/.env"]):
+        with patch.dict("os.environ", {}, clear=True), patch.object(
+            _em_mod, "_find_repo_root", return_value=tmp_path
+        ), patch.object(
+            _em_mod, "_SECONDARY_ENV_RELPATHS", ["argumentation_analysis/.env"]
+        ):
             with caplog.at_level(
                 logging.WARNING, logger="project_core.managers.environment_manager"
             ):
                 EnvironmentManager()
 
         assert not any(
-            "divergence" in record.message
-            for record in caplog.records
+            "divergence" in record.message for record in caplog.records
         ), "No WARNING when there is no secondary .env to diverge from"

--- a/tests/unit/test_1794_env_premise_guard.py
+++ b/tests/unit/test_1794_env_premise_guard.py
@@ -12,19 +12,16 @@ Measured by trap + identity probe in the same run: POST fired with
 Two guards:
 - source guard: the identity-swap motif is banned from the gate tree
   (it was red before the fix — 9 sites);
-- execution guard: after a predecessor enters AND EXITS a rebind (the
-  historical polluter, simulated here forever), the pl_2pass no-key scenario
-  emits zero watched-LLM request, observed by the #1787 counter.
+- execution guard: one deliberate swap must ENTER AND EXIT with the process
+  env object restored (identity, type, keys). The no-key-scenario variant
+  was removed: it measured the ambient mid-window writer (#1794 follow-up,
+  named by the os._Environ.__setitem__ tracer), not the swap mechanism —
+  it stayed red after the swap fix and conflated the two defects.
 """
 
-import asyncio
 import re
 from pathlib import Path
 from unittest.mock import patch
-
-import pytest
-
-import tests.llm_egress_counter as egress_module
 
 _TESTS_ROOT = Path(__file__).resolve().parents[1]
 
@@ -70,65 +67,24 @@ def test_no_os_environ_identity_swap_in_gate_tree():
     )
 
 
-def test_no_api_key_scenario_emits_nothing_after_a_rebind_roundtrip():
-    """Execution guard: a predecessor that entered and exited an os.environ
-    rebind (the historical polluter shape) must leave the no-key premise
-    enforceable — the #1787 counter must attribute zero watched-LLM request
-    to the pl_2pass no-key scenario run right after it."""
-    counter = egress_module.get_counter()
-    if counter is None:  # pragma: no cover - counter not active in this session
-        pytest.skip("LLM egress counter not active")
-
-    # The polluter's shape, kept on purpose: swap the process env, then let
-    # mock restore it. A leaking exit would leave os.environ a plain dict and
-    # the scenario below would fire (measured in the trap run of #1794).
-    sentinel = {"OPENROUTER_API_KEY": "sk-leak-should-not-survive-the-exit"}
-    with patch("os.environ", sentinel):
-        assert os_env_get("OPENROUTER_API_KEY") == "sk-leak-should-not-survive-the-exit"
-    assert (
-        os_env_get("OPENROUTER_API_KEY") != "sk-leak-should-not-survive-the-exit"
-    ), "the rebind outlived its with-block — process env is still the swapped dict"
-
-    from argumentation_analysis.core.shared_state import UnifiedAnalysisState
-    from argumentation_analysis.orchestration.invoke_callables import (
-        _invoke_propositional_logic,
-    )
-
-    state = UnifiedAnalysisState("Test argument for the guard scenario.")
-    ctx = {
-        "_state_object": state,
-        "source_metadata": {"opaque_id": "guard_1794"},
-        "arguments": ["National sovereignty requires immediate action"],
-    }
-
-    before = counter.total()
-    with patch.dict(
-        "os.environ",
-        {"OPENAI_API_KEY": "", "OPENROUTER_API_KEY": "", "OPENROUTER_BASE_URL": ""},
-    ):
-        asyncio.get_event_loop().run_until_complete(
-            _invoke_propositional_logic(state.raw_text, ctx)
-        )
-
-    attributed = [
-        r
-        for r in counter.snapshot()["requests"]
-        if r["test"].endswith(
-            "test_no_api_key_scenario_emits_nothing_after_a_rebind_roundtrip"
-        )
-        and r["class"] == "llm"
-    ]
-    assert not attributed, (
-        "the no-key scenario emitted watched-LLM request(s) after a clean "
-        "rebind roundtrip — the env premise is again decided outside the "
-        f"test: {attributed}"
-    )
-    assert (
-        counter.total() <= before + 1
-    )  # at most the counter's own control, never ours
-
-
-def os_env_get(key):
+def test_rebind_roundtrip_restores_process_environ():
+    """Execution guard: perform the historical polluter shape (one global
+    identity swap via patch) and assert the with-block EXIT restores the
+    process env object — same identity, still an _Environ, and the sentinel
+    key gone. A reintroduced leaking wrapper (swap that outlives its
+    with-block, the #1794 mechanism) fails here with no LLM call and no
+    counter dependency."""
     import os
 
-    return os.environ.get(key)
+    original = os.environ
+    sentinel = {"OPENROUTER_API_KEY": "sk-leak-should-not-survive-the-exit"}
+    with patch("os.environ", sentinel):
+        assert os.environ is sentinel, "patch('os.environ', ...) must swap the object"
+    assert os.environ is original, (
+        "the rebind outlived its with-block — process env is still the "
+        "swapped dict, so a later test's patch.dict window can resolve the "
+        "wrong object and leak provider keys (#1794)"
+    )
+    assert (
+        os.environ.get("OPENROUTER_API_KEY") != sentinel["OPENROUTER_API_KEY"]
+    ), "the sentinel key survived the exit"

--- a/tests/unit/test_1794_env_premise_guard.py
+++ b/tests/unit/test_1794_env_premise_guard.py
@@ -1,0 +1,134 @@
+"""#1794 — guards for the env-premise pollution family.
+
+The gate's two surviving real POSTs (#1591 residual) were caused by tests
+doing ``patch("<any-module>.os.environ", {...})``: mock resolves the module,
+then ``setattr(os, "environ", dict)`` — a GLOBAL identity swap, not a scoped
+one. During the swap window ``os.environ`` is a plain dict: dict writes never
+reach ``putenv`` (the C env block desynchronizes), and a later test's
+``patch.dict`` can resolve the swapped object instead of the process env.
+Measured by trap + identity probe in the same run: POST fired with
+``os.environ`` = ``builtins.dict``.
+
+Two guards:
+- source guard: the identity-swap motif is banned from the gate tree
+  (it was red before the fix — 9 sites);
+- execution guard: after a predecessor enters AND EXITS a rebind (the
+  historical polluter, simulated here forever), the pl_2pass no-key scenario
+  emits zero watched-LLM request, observed by the #1787 counter.
+"""
+
+import asyncio
+import re
+from pathlib import Path
+from unittest.mock import patch
+
+import pytest
+
+import tests.llm_egress_counter as egress_module
+
+_TESTS_ROOT = Path(__file__).resolve().parents[1]
+
+# patch("<module>.os.environ", ...), patch("os.environ", ...) and
+# patch.object(os, "environ", ...) all setattr the environ ATTRIBUTE of the
+# os module — a global identity swap. patch.dict(...) mutates the object and
+# is the only legitimate form.
+_SWAP_MOTIF = re.compile(
+    r"patch\(\s*[\"'](?:[\w.]+\.)?os\.environ[\"']"
+    r"|patch\.object\(\s*os\s*,\s*[\"']environ[\"']"
+)
+
+
+def _gate_test_files():
+    files = []
+    for sub in ("unit", "scripts"):
+        files.extend((Path(_TESTS_ROOT) / sub).rglob("*.py"))
+    return sorted(files)
+
+
+def test_no_os_environ_identity_swap_in_gate_tree():
+    """Source guard: patch("<mod>.os.environ", {...}) rebinds the process env
+    globally (#1794) — mutation via patch.dict(..., clear=True) is the only
+    legitimate form. Fails on any reintroduction of the swap motif."""
+    offenders = []
+    for f in _gate_test_files():
+        if f.name == Path(__file__).name:
+            # this guard deliberately performs one round-trip swap (execution
+            # guard below) — it asserts the restoration itself.
+            continue
+        try:
+            src = f.read_text(encoding="utf-8", errors="replace")
+        except OSError:
+            continue
+        for i, line in enumerate(src.splitlines(), start=1):
+            if _SWAP_MOTIF.search(line):
+                offenders.append(f"{f.relative_to(_TESTS_ROOT)}:{i}: {line.strip()}")
+    assert not offenders, (
+        "os.environ identity-swap patch(es) found — these rebind the process "
+        "env globally and leak provider keys into other tests' premises "
+        "(#1794). Convert to patch.dict('os.environ', {...}, clear=True):\n"
+        + "\n".join(offenders)
+    )
+
+
+def test_no_api_key_scenario_emits_nothing_after_a_rebind_roundtrip():
+    """Execution guard: a predecessor that entered and exited an os.environ
+    rebind (the historical polluter shape) must leave the no-key premise
+    enforceable — the #1787 counter must attribute zero watched-LLM request
+    to the pl_2pass no-key scenario run right after it."""
+    counter = egress_module.get_counter()
+    if counter is None:  # pragma: no cover - counter not active in this session
+        pytest.skip("LLM egress counter not active")
+
+    # The polluter's shape, kept on purpose: swap the process env, then let
+    # mock restore it. A leaking exit would leave os.environ a plain dict and
+    # the scenario below would fire (measured in the trap run of #1794).
+    sentinel = {"OPENROUTER_API_KEY": "sk-leak-should-not-survive-the-exit"}
+    with patch("os.environ", sentinel):
+        assert os_env_get("OPENROUTER_API_KEY") == "sk-leak-should-not-survive-the-exit"
+    assert (
+        os_env_get("OPENROUTER_API_KEY") != "sk-leak-should-not-survive-the-exit"
+    ), "the rebind outlived its with-block — process env is still the swapped dict"
+
+    from argumentation_analysis.core.shared_state import UnifiedAnalysisState
+    from argumentation_analysis.orchestration.invoke_callables import (
+        _invoke_propositional_logic,
+    )
+
+    state = UnifiedAnalysisState("Test argument for the guard scenario.")
+    ctx = {
+        "_state_object": state,
+        "source_metadata": {"opaque_id": "guard_1794"},
+        "arguments": ["National sovereignty requires immediate action"],
+    }
+
+    before = counter.total()
+    with patch.dict(
+        "os.environ",
+        {"OPENAI_API_KEY": "", "OPENROUTER_API_KEY": "", "OPENROUTER_BASE_URL": ""},
+    ):
+        asyncio.get_event_loop().run_until_complete(
+            _invoke_propositional_logic(state.raw_text, ctx)
+        )
+
+    attributed = [
+        r
+        for r in counter.snapshot()["requests"]
+        if r["test"].endswith(
+            "test_no_api_key_scenario_emits_nothing_after_a_rebind_roundtrip"
+        )
+        and r["class"] == "llm"
+    ]
+    assert not attributed, (
+        "the no-key scenario emitted watched-LLM request(s) after a clean "
+        "rebind roundtrip — the env premise is again decided outside the "
+        f"test: {attributed}"
+    )
+    assert (
+        counter.total() <= before + 1
+    )  # at most the counter's own control, never ours
+
+
+def os_env_get(key):
+    import os
+
+    return os.environ.get(key)

